File: algortihms-java/src/com/algo/tree/traversal/PostOrderTraversal.java
package com.algo.tree.traversal;

import com.algo.tree.Node;

import java.util.Stack;

public class PostOrderTraversal {
    public static void main(String[] args) {
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        Node node6 = new Node(6);
        node1.setLeft(node2);
        node1.setRight(node3);
        node2.setLeft(node4);
        node2.setRight(node5);
        node3.setLeft(node6);
        postOrder1(node1);
        System.out.println("\n");
        postOrder2(node1);
        System.out.println("\n");
        postOrder3(node1);
    }

    private static void postOrder1(Node root) {
        if (root == null) {
            return;
        }
        postOrder1(root.getLeft());
        postOrder1(root.getRight());
        System.out.print(root.getData() + "\t");
    }

    //Iterative traversal using 2 stacks
    private static void postOrder2(Node root) {
        if (root == null) {
            throw new IllegalArgumentException("Root cannot be null");
        }
        Stack<Node> stack1 = new Stack<>();
        Stack<Node> stack2 = new Stack<>();
        stack1.push(root);
        while (!stack1.isEmpty()) {
            Node node = stack1.pop();
            stack2.push(node);
            if (node.getLeft() != null) {
                stack1.push(node.getLeft());
            }
            if (node.getRight() != null) {
                stack1.push(node.getRight());
            }
        }
        while (!stack2.isEmpty()) {
            System.out.print(stack2.pop().getData() + "\t");
        }

    }

    //Iterative postorder traversal using 1 stack
    //1.Check if right child exist
    //2.if it exist, push to stack
    //3.push root to stack
    //4.pop element from stack
    //5.check if new top element of stack is right child of recently popped element
    //6.if yes pop it push root node in stack and set root to its right child
    //7.if right child in not new top element in stack then print it and set root to null
    private static void postOrder3(Node root) {
        if (root == null) {
            throw new IllegalArgumentException("Root of tree cannot be null");
        }
        Stack<Node> stack = new Stack<>();
        do {
            while (root != null) {
                if (root.getRight() != null) {
                    stack.push(root.getRight());
                }
                stack.push(root);
                root = root.getLeft();
            }
            root = stack.pop();
            if (!stack.isEmpty() && root.getRight() == stack.peek()) {
                stack.pop();
                stack.push(root);
                root = root.getRight();
            } else {
                System.out.print(root.getData() + "\t");
                root = null;
            }

        } while (!stack.isEmpty());


    }
}

File: algorithms-python/trees/traverse/postorder_traversal.py
"""
                        1
                    /      \
                    2        3
                  /   \\        \
                4        5       6
                                /
                               7
"""
from collections import deque
from typing import Optional, Deque

from trees.node import Node


def postorder3(root: Optional[Node]) -> None:
    if root is None:
        raise ValueError("Root element is none")
    stack: Deque[Node] = deque()
    while True:
        while root is not None:
            if root.right is not None:
                stack.append(root.right)
            stack.append(root)
            root = root.left
        if deque:
            root = stack.pop()
        # if top element of stack is right of popped element
        # then put root back and process right node
        if stack and root.right == stack[-1]:
            stack.pop()
            stack.append(root)
            root = root.right
        else:
            print(root.data, end="\t")
            root = None
        if not stack:
            break


# Iteration and using 2 stacks

def postorder2(root: Optional[Node]) -> None:
    if root is None:
        raise ValueError("Root is None")
    stack: Deque[Node] = deque()
    result: Deque[Node] = deque()
    stack.append(root)
    while stack:
        node: Node = stack.pop()
        result.append(node)
        if node.left is not None:
            stack.append(node.left)
        if node.right is not None:
            stack.append(node.right)
    while result:
        print(result.pop().data, end="\t")


# Recursion
def postorder1(root: Optional[Node]) -> None:
    if root is None:
        return
    postorder1(root.left)
    postorder1(root.right)
    print(root.data, end="\t")


def main():
    node1: Node = Node(1)
    node2: Node = Node(2)
    node3: Node = Node(3)
    node4: Node = Node(4)
    node5: Node = Node(5)
    node6: Node = Node(6)
    node7: Node = Node(7)
    node1.left = node2
    node1.right = node3
    node2.left = node4
    node2.right = node5
    node3.right = node6
    node6.left = node7
    postorder1(node1)
    print()
    postorder2(node1)
    print()
    postorder3(node1)


if __name__ == '__main__':
    main()
